File: AutoOptLib for Octave/Utilities/@DESIGN/Initialize.m
function [Operators,Paras] = Initialize(Setting,N)
% Initialize the designed algorithm(s).

%----------------------------Copyright-------------------------------------
% Copyright (C) <2023>  <Swarm Intelligence Lab>

% AutoOptLib is a free software. You can use, redistribute, and/or modify
% it under the terms of the GNU General Public License as published by the 
% Free Software Foundation, either version 3 of the License, or any later 
% version. 
%--------------------------------------------------------------------------

OpSpace   = Setting.OpSpace;
ParaSpace = Setting.ParaSpace;
Operators = cell(N,Setting.AlgP); % N algorithms, each with AlgP search 
% pathways. Each cell contains a number of AlgP matrix within edges of the graph 
% representation of a designed ealgorithm
Paras     = cell(N,1); % parameters of N algorithms
for i = 1:N
    % initialize operators
    indChoose = randi([OpSpace(1,1),OpSpace(1,2)]); % different search pathways use the same choose operator
    indUpdate = randi([OpSpace(3,1),OpSpace(3,2)]); % different search pathways use the same update operator
    for j = 1:Setting.AlgP
        currQ = randi(Setting.AlgQ); % number of search operators in the current search pathway
        Operators{i,j} = zeros(currQ+1,2);

        indSearch = randi([OpSpace(2,1),OpSpace(2,2)]);
        Operators{i,j}(1,:) = [indChoose,indSearch];

        indSearchStart = indSearch;
        for k = 2:currQ
            indSearchEnd = randi([OpSpace(2,1),OpSpace(2,2)]);
            Operators{i,j}(k,:) = [indSearchStart,indSearchEnd];
            indSearchStart = indSearchEnd;
        end

        Operators{i,j}(end,:) = [indSearchStart,indUpdate];
    end
    % initialize parameters
    tempPara = cell(length(ParaSpace),2); % first column of each row
    % contains a cloumn vector of an operator's parameter values,
    % second column will contain a string of whether the operator
    % performs local or global search behavior.
    indNonEmptyPara = find(~cellfun(@isempty,ParaSpace)==1);
    for j = 1:numel(indNonEmptyPara)
        k = indNonEmptyPara(j);
        tempPara{k,1} = ParaSpace{k}(:,1)+(ParaSpace{k}(:,2)-ParaSpace{k}(:,1)).*rand(size(ParaSpace{k},1),1);
    end
    Paras{i} = tempPara;
end
end

File: src/autooptlib/utils/design/_initialize.py
"""Python port of the MATLAB Initialize routine."""

from __future__ import annotations

from typing import Any

import numpy as np

from ._helpers import ensure_rng, get_flex


def initialize(setting: Any, n: int):
    """Initialize the designed algorithm(s)."""
    alg_count = int(n)
    if alg_count <= 0:
        return [], []

    rng = ensure_rng(setting)
    op_space = np.array(get_flex(setting, "op_space", required=True), dtype=int)
    para_space = list(get_flex(setting, "para_space", required=True))
    alg_p = int(get_flex(setting, "alg_p", required=True))
    alg_q = int(get_flex(setting, "alg_q", required=True))

    operators = [[None for _ in range(alg_p)] for _ in range(alg_count)]
    paras: list[list[list[Any]]] = [[None for _ in range(len(para_space))] for _ in range(alg_count)]

    for i in range(alg_count):
        ind_choose = int(rng.integers(op_space[0, 0], op_space[0, 1] + 1))
        ind_update = int(rng.integers(op_space[2, 0], op_space[2, 1] + 1))
        for j in range(alg_p):
            curr_q = int(rng.integers(1, alg_q + 1))
            matrix = np.zeros((curr_q + 1, 2), dtype=int)
            ind_search = int(rng.integers(op_space[1, 0], op_space[1, 1] + 1))
            matrix[0, :] = (ind_choose, ind_search)
            ind_start = ind_search
            for k in range(1, curr_q):
                ind_end = int(rng.integers(op_space[1, 0], op_space[1, 1] + 1))
                matrix[k, :] = (ind_start, ind_end)
                ind_start = ind_end
            matrix[-1, :] = (ind_start, ind_update)
            operators[i][j] = matrix

        ind_non_empty_para = [idx + 1 for idx, space in enumerate(para_space) if space is not None and len(space) > 0]
        temp_para = [[None, None] for _ in range(len(para_space))]
        for idx in ind_non_empty_para:
            bounds = np.asarray(para_space[idx - 1], dtype=float)
            lower = bounds[:, 0]
            upper = bounds[:, 1]
            values = lower + (upper - lower) * rng.random(lower.shape)
            temp_para[idx - 1][0] = values
        paras[i] = temp_para

    return operators, paras
